planner/src/planner.cpp

```diff
@@ -0,0 +1,333 @@
+}
+
+void fsm::TakeOff(CmdTakeOff const& cmd) {
+	ros::Rate rate(20);
+	ROS_INFO("Taking off!");
+
+	ROS_WARN("Waiting for connection..");
+	while (ros::ok() && current_state_.connected == false) {
+		ros::spinOnce();
+		rate.sleep();
+	}
+	ROS_INFO("Connected to MAVROS..");
+
+	geometry_msgs::PoseStamped pose;
+	pose.pose.position.x = 0;
+	pose.pose.position.y = 0;
+	pose.pose.position.z = 2.5;
+
+	for (int i = 100; ros::ok() && i > 0; --i) {
+		pose_pub_.publish(pose);
+		ros::spinOnce();
+		rate.sleep();
+	}
+	ROS_WARN("Published setpoint 100 times!");
+
+	mavros_msgs::SetMode offb_set_mode;
+	offb_set_mode.request.custom_mode = "OFFBOARD";
+
+	mavros_msgs::CommandBool arm_cmd;
+	arm_cmd.request.value = true;
+
+	while (!arm_cmd.response.success) {
+		arming_client_.call(arm_cmd);
+		ROS_INFO_ONCE("Arming...");
+	}
+
+	ros::Time last_request = ros::Time::now();
+
+	while (ros::ok()) {
+		if (odom_.pose.pose.position.z >= 1.8) {
+			offb_set_mode.request.custom_mode = "AUTO.LOITER";
+			set_mode_client_.call(offb_set_mode);
+			ROS_INFO_ONCE("AUTO.LOITER Triggered");
+			break;
+		}
+
+		if (current_state_.mode != "OFFBOARD" && (ros::Time::now() - last_request > ros::Duration(5.0))) {
+			if (set_mode_client_.call(offb_set_mode)) {
+				if (offb_set_mode.response.mode_sent) ROS_WARN("Offboard enabled");
+			}
+			last_request = ros::Time::now();
+		} else {
+			if (!current_state_.armed && (ros::Time::now() - last_request > ros::Duration(5.0))) {
+				if (arming_client_.call(arm_cmd)) {
+					if (arm_cmd.response.success) ROS_WARN("Vehicle armed");
+				}
+				last_request = ros::Time::now();
+			}
+		}
+		pose_pub_.publish(pose);
+		ros::spinOnce();
+		rate.sleep();
+	}
+}
+
+void fsm::DetectionBased(CmdEstimated const& cmd) {
+	ROS_INFO("Entering detection based event");
+
+	mavros_msgs::SetMode offb_set_mode;
+	offb_set_mode.request.custom_mode = "OFFBOARD";
+	set_mode_client_.call(offb_set_mode);
+
+	ros::Rate looprate(20);
+	bool flag = true;
+	float sum_x = 0, sum_y = 0, sum_z = 0;
+	int k = 1;
+
+	while (flag) {
+		ros::spinOnce();
+
+		if (!(centre_.x == -1 || centre_.y == -1) && k == 1) {
+			int c = 5;
+			while (c > 0) {
+				ros::spinOnce();
+				if (!(centre_.x == -1 || centre_.y == -1)) {
+					sum_x += estimated_pose_.x;
+					sum_y += estimated_pose_.y;
+					sum_z += estimated_pose_.z;
+					c--;
+				} else {
+					std::cout << "centre is not being detected" << std::endl;
+				}
+			}
+
+			frame_vec_(0) = sum_x / 5;
+			frame_vec_(1) = sum_y / 5;
+			frame_vec_(2) = sum_z / 5;
+			std::cout << frame_vec_(0) << " " << frame_vec_(1) << " " << frame_vec_(2) << std::endl;
+			k--;
+		}
+
+		double d = centre_.d;
+		double x = odom_.pose.pose.position.x;
+		double y = odom_.pose.pose.position.y;
+		double z = odom_.pose.pose.position.z;
+
+		if ((centre_.x == -1 || centre_.y == -1) && (sum_x == 0 && sum_y == 0 && sum_z == 0)) {
+			double dist = sqrt((sq(rough_pose_[p % 4][0] - x)) + (sq(rough_pose_[p % 4][1] - y)));
```

Comment:
use `rough_pose_.size()` instead of hardcoding 4. then it will work for any number of frames